Add support for NeMo

titu1994 opened the pull request:
# Changelog
- Add support for NeMo API inference

Narsil commented on docker_images/nemo/app/pipelines/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,74 @@
+        
+        # Pre-Initialize RNNT decoding strategy
+        if hasattr(self.model, 'change_decoding_strategy'):
+            model.change_decoding_strategy(None)
```
```suggestion
            self.model.change_decoding_strategy(None)
```

Narsil reviewed (approved):
Awesome PR ! 

The code seemed to work in my env

```
./manage.py start --model-id=nvidia/stt_en_conformer_ctc_medium --task automatic-speech-recognition --framework nemo
```

However the docker image seems to fail:
```
./manage.py docker --model-id=nvidia/stt_en_conformer_ctc_medium --task automatic-speech-recognition --framework nemo
```

With 

```
  File "/usr/local/lib/python3.8/site-packages/torchmetrics/functional/audio/__init__.py", line 26, in <module>
    from torchmetrics.functional.audio.pesq import perceptual_evaluation_speech_quality  # noqa: F401
  File "/usr/local/lib/python3.8/site-packages/torchmetrics/functional/audio/pesq.py", line 20, in <module>
    import pesq as pesq_backend
  File "/usr/local/lib/python3.8/site-packages/pesq/__init__.py", line 6, in <module>
    from .cypesq import cypesq, cypesq_retvals, cypesq_error_message as pesq_error_message
  File "pesq/cypesq.pyx", line 1, in init cypesq
ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject
```

Google led me here : https://githubhot.com/repo/PyTorchLightning/metrics/issues/841
Any idea where that might come from ?

Is it possible to add another docker tests ?
```diff
diff --git a/tests/test_dockers.py b/tests/test_dockers.py
index bafd22e..9985518 100644
--- a/tests/test_dockers.py
+++ b/tests/test_dockers.py
@@ -273,6 +273,13 @@ class DockerImageTests(unittest.TestCase):
             dataset_column="file",
         )

+    def test_nemo(self):
+        self.framework_docker_test(
+            "nemo",
+            "automatic-speech-recognition",
+            "nvidia/stt_en_conformer_ctc_medium",
+        )
+
```

We will add it in a later PR if it's inconvenient.

- The current model under test is private, is it possible to make it public ? (the API community does not work for private models for now)
- Can we edit the README.md on the model page to add:
  ` - library_name: nemo` ? This will be taken care of later when we modify the JS part as it will get inferred automatically from tags, this is a simple override that allows to not wait for this.
  
  This still needs a small modification on the API (registering the new `nemo` framework) which should some shortly after this PR is passed ! ).

Narsil commented:
For `quality` you probably just need to run

`make style` This should fix everything.

titu1994 commented on docker_images/nemo/app/pipelines/automatic_speech_recognition.py:
```diff
@@ -0,0 +1,74 @@
+            audio_path = os.path.join(tmpdir, f'audio_{uuid.uuid4()}.wav')
+            soundfile.write(audio_path, inputs, self.sampling_rate)
+
+            transcriptions = self.model.transcribe([audio_path])
```
The NeMo ASR pipeline is quite involved, and we have a layered API to perform inference. transcribe() kind of generalizes the API across various scopes (5 different encoders x {ctc decoder, rnnt decoder} x {char, subword} x languages).  Since it is designed for offline transcription of files, it somewhat depends on the assumption that file paths are provided so that the data loader can be reused per model. 

The audio is stored temporarily (to reduce data storage concerns) and is immediately deleted after inference. Its definitely a performance hit, perhaps later nemo versions can have higher level APIs that deal with audio directly. Once we stabilize such an API, we can update here. Does that seem reasonable?

Narsil replied:
Very reasonable ! what I suggest is we keep the code as-is, we check real world performance. If it's not too bad we keep it as-is, if it is I will enable some kind of `tmpfs`.

When there's better code we can update as desired.

titu1994 commented:
> - library_name: nemo ? This will be taken care of later when we modify the JS part as it will get inferred automatically from tags, this is a simple override that allows to not wait for this.

Should this be part of the `tags`? Or a separate argument outside it?

titu1994 commented:
Addressed most of the comments + added the docker test. It wont pass since the model is currently private, but I'll try to get a signoff for the medium model for testing purposes.

Narsil reviewed (approved):
LGTM.

I can fix the style for you if you're ok with it.
I think it's a black version issue.

titu1994 commented:
Yep please fix style. I did make locally but it seems to have no effect, simply prints a warning (but does do isort and black).
